### PR TITLE
[experiment] Don't relate primitives to their wrappers

### PR DESCRIPTION
**Pull request by jakebailey**
Basically, I was trying to understand why we allow `const x: Number = 1` when the latter is a wrapper object. e.g. `1 === new Number(1)` will yield false and `typeof new Number(1) === "object"`, not `number`.

@DanielRosenwasser said this might be an interesting experiment. Might as well try.

### CODE REVIEW

**Comment by jakebailey** [comment]
@typescript-bot test this
@typescript-bot test top100
@typescript-bot user test this
@typescript-bot run dt
@typescript-bot perf test this faster
@typescript-bot pack this

**Comment by DanielRosenwasser** [comment]
> why we allow `const x: number = new Number(1)`

It's the opposite - `number` is assignable (and comparable) to `Number`; `Number` is not assignable (nor comparable) to `number`.

```ts
const a: number = new Number(0); // error

const b: Number = 0; // allowed - structurally compatible
```

**Comment by jakebailey** [comment]
> > why we allow `const x: number = new Number(1)`
> 
> It's the opposite - `number` is assignable (and comparable) to `Number`; `Number` is not assignable (nor comparable) to `number`.
> 
> ```ts
> const a: number = new Number(0); // error
> 
> const b: Number = 0; // allowed - structurally compatible
> ```

Yep, sorry. Flipped it when writing!

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -21570,8 +21574,13 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                 }
                 const sourceIsPrimitive = !!(sourceFlags & TypeFlags.Primitive);
                 if (relation !== identityRelation) {
+                    const originalSource = source
+                    const originalFlags = sourceFlags;
                     source = getApparentType(source);
                     sourceFlags = source.flags;
+                    if (source !== originalSource && originalFlags & TypeFlags.HasWrapper && !(sourceFlags & TypeFlags.HasWrapper) && source === target) {
+                        return Ternary.False;
+                    }
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts:21583]: I actually am not sure if you want to disallow primitives from being assignable and comparable to wrapper objects. I think there's at least some rationale that you actually want to allow `1 as Number` even if that is a bit silly - but I can't imagine why.

Given that, I think the right place for this is an ad-hoc check in `===` and `!==`, and possibly `==` and `!=` too. But this is fine for prototyping.

Comment (jakebailey) [reply src/compiler/checker.ts:21583]: That's true, I think I misinterpreted your suggestion.

Comment (DanielRosenwasser) [reply src/compiler/checker.ts:21583]: It's not the craziest idea - we highly discourage using the wrapper objects anyway.

**Comment by Andarist** [comment]
driveby q: what does "comparable" really mean?

**Comment by DanielRosenwasser** [comment]
Comparability is a relaxed form of assignability. It aims to answer "is this plausibly of the same type" instead of "is this pretty much substitutable".

See #17214 for an oldish list of differences between comparability/assignability.

**Comment by jakebailey** [comment]
Fwiw I really like the result here. I almost think we should re-alloe comparing number and Number before 5.0 goes out, and then do _this_ PR instead.

**Comment by jakebailey** [comment]
@typescript-bot pack this

**Comment by RyanCavanaugh** [comment]
The cure might worse than the disease here. I'm nervous that this introduces subtyping commutativity violations, and does so nominally.
```
{ length: number } <-- String <-/- string
```
```
{ length: number } <-- string
```
`string` would be assignable to e.g. `Pick<String, keyof String>` but not `String` ?

**Comment by jakebailey** [comment]
I should disclaim that this is definitely not the final iteration of this PR, but something I threw together late at night.

Probably we could just do the conservative thing of not allowing `string === String` like we do `number === NaN`. That, or some other thing to make this better.

**Comment by jakebailey** [comment]
@typescript-bot pack this

**Comment by jakebailey** [comment]
@typescript-bot pack this

**Comment by fatcerberus** [comment]
I always thought the reason `number` is assignable to/comparable with `Number` was structural — the same reason primitives are assignable to `{}` and more to the point, e.g. `{ toString(): string }`.

**Comment by jakebailey** [comment]
Yep, that's right (what Ryan was saying above). In this case, we get the "apparent" type which is the wrapper and has all of the properties and methods, and that's structurally compared.

**Comment by jakebailey** [comment]
I'm going to close this as this isn't the kind of approach we'd be taking. It also didn't seem to catch any bugs, just people using `Number` as a type annotation, which is of course not my favorite, but not actually bug-causing it seems.

I might come back to doing this like `NaN` and try to error when `Number` is compared with `number`, but that would be a different PR entirely.

**Pull request closed without merging**